src/backend/distributed/commands/multi_copy.c

```diff
@@ -0,0 +1,1177 @@
+
+	if (partitionMethod == DISTRIBUTE_BY_HASH)
+	{
+		int hashedValue = DatumGetInt32(FunctionCall1(hashFunction,
+													  partitionColumnValue));
+		if (useBinarySearch)
+		{
+			shardInterval = SearchCachedShardInterval(Int32GetDatum(hashedValue),
+													  shardIntervalCache, shardCount,
+													  compareFunction);
+		}
+		else
+		{
+			uint64 hashTokenIncrement = HASH_TOKEN_COUNT / shardCount;
+			int shardIndex = (uint32) (hashedValue - INT32_MIN) / hashTokenIncrement;
+
+			shardInterval = shardIntervalCache[shardIndex];
+		}
+	}
+	else
+	{
+		shardInterval = SearchCachedShardInterval(partitionColumnValue,
+												  shardIntervalCache, shardCount,
+												  compareFunction);
+	}
+
+	return shardInterval;
+}
+
+
+/*
+ * SearchCachedShardInterval performs a binary search for a shard interval matching a
+ * given partition column value and returns it.
+ */
+static ShardInterval *
+SearchCachedShardInterval(Datum partitionColumnValue, ShardInterval **shardIntervalCache,
+						  int shardCount, FmgrInfo *compareFunction)
+{
+	int lowerBoundIndex = 0;
+	int upperBoundIndex = shardCount;
+
+	while (lowerBoundIndex < upperBoundIndex)
+	{
+		int middleIndex = (lowerBoundIndex + upperBoundIndex) >> 1;
+		int maxValueComparison = 0;
+		int minValueComparison = 0;
+
+		minValueComparison = FunctionCall2Coll(compareFunction,
+											   DEFAULT_COLLATION_OID,
+											   partitionColumnValue,
+											   shardIntervalCache[middleIndex]->minValue);
+
+		if (DatumGetInt32(minValueComparison) < 0)
+		{
+			upperBoundIndex = middleIndex;
+			continue;
+		}
+
+		maxValueComparison = FunctionCall2Coll(compareFunction,
+											   DEFAULT_COLLATION_OID,
+											   partitionColumnValue,
+											   shardIntervalCache[middleIndex]->maxValue);
+
+		if (DatumGetInt32(maxValueComparison) <= 0)
+		{
+			return shardIntervalCache[middleIndex];
+		}
+
+		lowerBoundIndex = middleIndex + 1;
+	}
+
+	return NULL;
+}
+
+
+/*
+ * OpenCopyTransactions opens a connection for each placement of a shard and
+ * starts a COPY transaction. If a connection cannot be opened, then the shard
+ * placement is marked as inactive and the COPY continues with the remaining
+ * shard placements.
+ */
+static void
+OpenCopyTransactions(CopyStmt *copyStatement, ShardConnections *shardConnections,
+					 int64 shardId)
+{
+	List *finalizedPlacementList = NIL;
+	List *failedPlacementList = NIL;
+	ListCell *placementCell = NULL;
+	ListCell *failedPlacementCell = NULL;
+	List *connectionList = NIL;
+
+	finalizedPlacementList = FinalizedShardPlacementList(shardId);
+
+	foreach(placementCell, finalizedPlacementList)
+	{
+		ShardPlacement *placement = (ShardPlacement *) lfirst(placementCell);
+		char *nodeName = placement->nodeName;
+		int nodePort = placement->nodePort;
+		TransactionConnection *transactionConnection = NULL;
+		StringInfo copyCommand = NULL;
+		PGresult *result = NULL;
+
+		PGconn *connection = ConnectToNode(nodeName, nodePort);
```

Comment:
Mainly because it was originally written for pg_shard, but it also kept the code simpler.